### ISSUE TITLE
Deadlock during Resolve

### ISSUE THREAD

**Issue by Artemigos** [dadhi/DryIoc#391]
I've run into a weird deadlock when resolving one of the services manually.

First short background - we're using DryIoc in a WPF app, and it's working very nicely for us right now. As part of some concept work I'm trying to run the code as a `netcoreapp3.1` console app. All of our registration code is being reused for this, so the container contents are very similar to what we have in WPF. The app resolves and configures some things and then one of the resolves deadlocks. I updated to 4.7.5 and it still happens, the call stack looks like this:

```
System.Private.CoreLib.dll!System.Threading.SpinWait.SpinOnceCore(int sleep1Threshold) Line 197	C#
DryIoc.dll!DryIoc.Scope.WaitForItemIsSet(ImTools.ImMapEntry<object> itemRef) Line 12019	C#
DryIoc.dll!DryIoc.Scope.TryGetOrAddWithoutClosure(int id, DryIoc.IResolverContext resolveContext, FastExpressionCompiler.LightExpression.Expression expr, bool useFec, System.Func<DryIoc.IResolverContext, FastExpressionCompiler.LightExpression.Expression, bool, object> createValue, int disposalOrder) Line 12109	C#
DryIoc.dll!DryIoc.Interpreter.TryInterpretMethodCall(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.Expression expr, object paramExprs, object paramValues, DryIoc.Interpreter.ParentLambdaArgs parentArgs, bool useFec, ref object result) Line 3622	C#
DryIoc.dll!DryIoc.Interpreter.TryInterpret(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.Expression expr, object paramExprs, object paramValues, DryIoc.Interpreter.ParentLambdaArgs parentArgs, bool useFec, out object result) Line 3023	C#
DryIoc.dll!DryIoc.Interpreter.TryInterpretAndUnwrapContainerException(DryIoc.IResolverContext r, FastExpressionCompiler.LightExpression.Expression expr, bool useFec, out object result) Line 2864	C#
DryIoc.dll!DryIoc.Container.ResolveAndCache(int serviceTypeHash, System.Type serviceType, DryIoc.IfUnresolved ifUnresolved) Line 409	C#
DryIoc.dll!DryIoc.Container.DryIoc.IResolver.Resolve(System.Type serviceType, DryIoc.IfUnresolved ifUnresolved) Line 356	C#
DryIoc.dll!DryIoc.Resolver.Resolve(DryIoc.IResolver resolver, System.Type serviceType) Line 8008	C#
```

I was trying to figure out what's causing this so I started reading DryIoc code and one method in particular looks suspicious:

![image](https://user-images.githubusercontent.com/4130682/113974618-86174b80-983e-11eb-90cb-33db22d883b5.png)

The deadlock happens inside the `WaitForItemIsSet` method and I don't understand how it's ever going to get set. Is this maybe a bug?

**Comment by dadhi** [comment]
@Artemigos Got it. I will check.

**Comment by dadhi** [comment]
@Artemigos How many threads are accessing this part? 
Is it just the UI and background thread?

Do you have some recursive dependencies, I mean by design?

The only thing I can imagine (now) to deadlock on this wait is the one thread failing to resolve scoped dependecy for any reason, so not setting the `otherItemRef` into any value. Fixing the reason for this fail should fix the deadlock.

**Comment by Artemigos** [comment]
This is happening in a console app but with `async Task Main()` - the Resolve runs on a different thread already, because it's a continuation after a few awaits. The main thread is blocked waiting for the tasks to finish. I'm not actually doing anything in parallel, so exactly one thing is hitting DryIoc. Also the deadlock is consistent - happens every time.

There are no recursive dependencies, we had one case of that, but we solved it long time ago by doing a lazy resolve (manually).

I downgraded do 4.1.4 and with this version I'm getting an exception - one of the dependencies can't be resolved. I'm not sure why this is happening yet, I'll let you know what I find.

Also, I'm curious - where is the `Value` of the `otherItemRef` being set?

**Comment by dadhi** [comment]
> I downgraded do 4.1.4 and with this version I'm getting an exception

This is already a good finding.

> where is the Value of the otherItemRef being set?

By other thread. But you're saying you don't have one so it is suspicious. Will be staring more.. and I need to introduce some diagnostics for such cases.

**Comment by Artemigos** [comment]
I was missing a project reference, so the dependency was actually missing. After upgrading to 4.7.5 again it works, so I can solve my problem now. The only thing left is the deadlock when things fail. I'll take a look at the threads a little more, until now I was looking at the "Parallel Stacks" window in VS, which seems to be a simplified view with some special Task handling.

**Comment by Artemigos** [comment]
There are no other threads:

![image](https://user-images.githubusercontent.com/4130682/113997176-2e85d980-9858-11eb-9459-59b77cff3482.png)

`Main Thread` - `Main()` waiting for tasks
`GC Finalizer Thread` - not relevant for us
worker thread in `SpinWait` - this one is deadlocked

**Comment by Artemigos** [comment]
Tomorrow I might try to reproduce this on some small sample, but it's hard when I don't know what the actual cause is. The weirdest thing we do that I can think of is registering a lot of singletons with `asResolutionCall` set to `true` and then the interfaces are a mapping to that singleton.

What is spawning the thread that should be filling the `Value` that we talked about before?

**Comment by dadhi** [comment]
> What is spawning the thread that should be filling the Value that we talked about before?

Nothing from the DryIoc. 
`asResolutionCall` is the interesting case. Are you using many reconfigurable dependencies?

**Comment by dadhi** [comment]
@Artemigos 

Here is how the single-threaded recursive dependrency may cause the deadlock, but it should be prevented by the DryIoc check prior to the Resolve. So the cause maybe the check is not working in certain conditions.

![P10408-205919](https://user-images.githubusercontent.com/39516/114075095-d11c7780-98ad-11eb-97fb-cc80f47c44f5.jpg)

**Comment by Artemigos** [comment]
By "manual lazy resolve" I just mean that we break the circular chain by doing a `Resolve`  when the service is actually needed instead of injecting into the constructor.

As for the `asResolutionCall` - we use it almost everywhere. This codebase was using a different container before and to get DryIoc to work with existing code I had to think a little out of the box.

I don't think that a circular dependency is causing this here. After I added the missing project reference and everything got registered, the resolve works as intended, so the error was a missing registration (we have this automated based on a convention, that's why a missing reference can fail silently). I can actually prove this - when I move the `Resolve` call a few lines earlier in the program I get the exception saying that one of the dependencies cannot be resolved.

I'm trying to narrow this down, but I still can't reproduce this in an isolated console project. For now I have this, but this still works "as intended" (throws an exception):

```csharp
using System;
using System.Threading.Tasks;
using DryIoc;

namespace ConsoleApp19
{
    public interface IA {}
    public interface IB {}
    public interface IC {}

    public class A : IA
    {
        private IB b;
        public A(IB b) => this.b = b;
    }

    public class C : IC {}

    class Program
    {
        static async Task Main(string[] args)
        {
            var container = new Container(rules => rules
                .With(FactoryMethod.ConstructorWithResolvableArguments)
                .WithoutEagerCachingSingletonForFasterAccess()
                .WithoutThrowOnRegisteringDisposableTransient()
                .WithDefaultIfAlreadyRegistered(IfAlreadyRegistered.Replace));

            container.Register(typeof(A), Reuse.Singleton, setup: Setup.With(asResolutionCall: true));
            container.RegisterMany(new [] { typeof(A), typeof(IA) }, typeof(A), Reuse.Singleton, setup: Setup.With(asResolutionCall: true));
            container.Register(typeof(C), Reuse.Singleton, setup: Setup.With(asResolutionCall: true));
            container.RegisterMany(new [] { typeof(C), typeof(IC) }, typeof(C), Reuse.Singleton, setup: Setup.With(asResolutionCall: true));

            var c = container.Resolve<IC>();

            await Task.Yield();

            var a = container.Resolve<IA>();
        }
    }
}
```

Maybe this is already a hint for you.

BTW I know it's not necessary to do a registration before the `RegisterMany` - I'm just trying to mimick what our automated registrations end up doing.

**Comment by dadhi** [comment]
@Artemigos Thanks for the code. From the first glance, it is nothing suspicious but will be digging further.

**Comment by Artemigos** [comment]
Yup, the deadlock is solved for me - providing all the dependencies got rid of the problem.

**Comment by Artemigos** [comment]
Found it:

```csharp
using System;
using DryIoc;

namespace ConsoleApp19
{
    // A -> B -> C -> D(missing)
    //   \----->
    public interface IA {}
    public interface IB {}
    public interface IC {}
    public interface ID {}

    public class A : IA
    {
        private IB b;
        private IC c;

        public A(IC c,IB b)
        {
            this.b = b;
            this.c = c;
        }
    }

    public class B : IB
    {
        private IC c;
        public B(IC c) => this.c = c;
    }

    public class C : IC
    {
        private ID d;
        public C(ID d) => this.d = d;
    }

    class Program
    {
        static void Main(string[] args)
        {
            var container = new Container(rules => rules
                .With(FactoryMethod.ConstructorWithResolvableArguments)
                .WithoutEagerCachingSingletonForFasterAccess()
                .WithoutThrowOnRegisteringDisposableTransient()
                .WithDefaultIfAlreadyRegistered(IfAlreadyRegistered.Replace));

            container.Register(typeof(A), Reuse.Singleton, setup: Setup.With(asResolutionCall: true));
            container.RegisterMany(new [] { typeof(A), typeof(IA) }, typeof(A), Reuse.Singleton, setup: Setup.With(asResolutionCall: true));
            container.Register(typeof(B), Reuse.Singleton, setup: Setup.With(asResolutionCall: true));
            container.RegisterMany(new [] { typeof(B), typeof(IB) }, typeof(B), Reuse.Singleton, setup: Setup.With(asResolutionCall: true));
            container.Register(typeof(C), Reuse.Singleton, setup: Setup.With(asResolutionCall: true));
            container.RegisterMany(new [] { typeof(C), typeof(IC) }, typeof(C), Reuse.Singleton, setup: Setup.With(asResolutionCall: true));

            Resolve(container);
            Resolve(container);
        }

        static void Resolve(IContainer container)
        {
            try
            {
                container.Resolve<IA>();
            }
            catch (ContainerException e)
            {
                Console.WriteLine(e);
            }
        }
    }
}
```

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="DryIoc.dll" Version="4.7.5" />
  </ItemGroup>

</Project>
```

It deadlocks on the second resolve. The dependency structure is also important.

**Comment by dadhi** [comment]
@Artemigos Great, I will get it from here.

**Comment by dadhi** [comment]
@Artemigos 

I have provided the fix introducing the waiting timeout because at the moment it is "the better than nothing" solution and maybe actually enough for everyone. Hope that the error message below will provide the clues for the problem reasons

```
WaitForScopedServiceIsCreatedTimeoutExpired = Of(
    "We have waited for the creation of the scoped service by 'other party' for the {0} ticks without the success." + NewLine +
    "It means that either the 'other party' is the parallel thread which has started! but unable to finish the creation of the service in the provided amount of time." + NewLine +
    "Or more likely we have the undetected recursive dependency and the 'other party' is the same thread." + NewLine +
    "Another reason may be that the previous scoped service resolution is failed with the exception and the exception was CATCHED " + NewLine +
    "but you are proceeding to resolve the failed service again." + NewLine + 
    "That's why for all these reasons we have a timeout to prevent the hanging due to the infinite waiting.");

```

I am open to ideas of error handling and its message improvements.

**Comment by Artemigos** [comment]
The message looks fine - it's much more informative than a deadlock :)

When doing a "proper" fix I'd probably take inspiration from here: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Lazy.cs

It seems that exception is also considered a "finished" state for a resolution and I'd suspect that all potential threads will get the same exception. This would probably require some changes deep in DryIoc's thread-safety code, so I'm not sure if that's something you want to do, though.

**Comment by dadhi** [comment]
> When doing a "proper" fix I'd probably take inspiration from here: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Lazy.cs

Thanks for the idea. I will probably put it on hold until I get more evidence of the timeout inefficiency.

**Comment by dadhi** [comment]
@Artemigos I will close it for now but will link the issue from the discussion to not forget.
